rio_tiler/reader.py

```diff
@@ -46,129 +85,182 @@ def read(
-            warnings.warn(
-                "Alpha band was removed from the output data array", AlphaBandWarning
-            )
+    if max_size and width and height:
+        warnings.warn(
+            "'max_size' will be ignored with with 'height' and 'width' set.",
+            UserWarning,
+        )
 
-    band_names = [f"b{idx}" for idx in indexes]
+    resampling = Resampling[resampling_method]
+    dst_crs = dst_crs or src_dst.crs
+    with contextlib.ExitStack() as ctx:
+        # Use WarpedVRT when Re-projection or Nodata or User VRT Option (cutline)
+        if (dst_crs != src_dst.crs) or nodata is not None or vrt_options:
+            vrt_params = {
+                "crs": dst_crs,
+                "add_alpha": True,
+                "resampling": resampling,
+            }
 
-    vrt_params = dict(add_alpha=True, resampling=Resampling[resampling_method])
-    nodata = nodata if nodata is not None else src_dst.nodata
-    if nodata is not None:
-        vrt_params.update(dict(nodata=nodata, add_alpha=False, src_nodata=nodata))
+            nodata = nodata if nodata is not None else src_dst.nodata
+            if nodata is not None:
+                vrt_params.update(
+                    {"nodata": nodata, "add_alpha": False, "src_nodata": nodata}
+                )
 
-    if has_alpha_band(src_dst):
-        vrt_params.update(dict(add_alpha=False))
+            if has_alpha_band(src_dst):
+                vrt_params.update({"add_alpha": False})
 
-    if vrt_options:
-        vrt_params.update(vrt_options)
+            if vrt_options:
+                vrt_params.update(**vrt_options)
 
-    with WarpedVRT(src_dst, **vrt_params) as vrt:
-        if not window:
-            window = windows.Window(
-                col_off=0, row_off=0, width=vrt.width, height=vrt.height
-            )
+            # TODO: Check if we fetch the Overviews when not using transform
+            dataset = ctx.enter_context(WarpedVRT(src_dst, **vrt_params))
 
-        if ColorInterp.alpha in vrt.colorinterp:
-            indexes = tuple(indexes) + (vrt.colorinterp.index(ColorInterp.alpha) + 1,)
-            data = vrt.read(
-                indexes=indexes,
+        else:
+            dataset = src_dst
+
+        if max_size and not (width and height):
+            height, width = _get_width_height(max_size, dataset.height, dataset.width)
+
+        if indexes is None:
+            indexes = non_alpha_indexes(dataset)
+
+        boundless = False
+        if window:
+            if isinstance(window, tuple):
+                window = windows.Window.from_slices(
+                    *window, height=dataset.height, width=dataset.width, boundless=True
+                )
+            if min(window.col_off, window.row_off) < 0:
+                boundless = True
+
+            wbounds = dataset.window_bounds(window)
+            if wbounds[2] > dataset.bounds[2] or wbounds[3] > dataset.bounds[3]:
+                boundless = True
```

Comment:
boundless is not a allowed for warpedVRT so we can't set it to True automatically. (we could maybe set it to True and then set it to false for WarpedVRT 🤷